src/ast/ddl.rs

```diff
@@ -46,12 +47,12 @@ pub enum AlterTableOperation {
     /// `DROP CONSTRAINT [ IF EXISTS ] <name>`
     DropConstraint {
         if_exists: bool,
-        name: Ident,
+        name: Located<Ident>,
```

Comment:
What is the purpose of wrapping this in `Located<>`? 

Another alternative would be to add a location field to the `Ident` struct like

```rust
pub struct Ident {
    /// The value of the identifier without quotes.
    pub value: String,
    /// The starting quote if any. Valid quote characters are the single quote,
    /// double quote, backtick, and opening square bracket.
    pub quote_style: Option<char>,
    /// location in the source tree. <----- New Field
    location: Location,
}
```

Both will be breaking changes, but I am thinking to a future where there is location information on all structs -- if all fields all wrapped in `Location<..>` it is going to be a lot of replicated `Location<>` in the code

If they each have a `location` field then the location will only be added by the parser (though the tests will need updating too)

Reply:
Great question. For some background context, I've been working on a SQL compiler that uses the `sqlparser-rs` library as our parser. We added these locations to improve the error messages in the compiler.

We actually started by embedding the `location` in `Ident`, but there are places where we actually want to store the `Ident` itself (e.g. a map from `Ident` to some binder state about it), and in those places we don't care about the `Location`. There are also places where we want to do things like convert a `string` into an `Ident`, and if `location` is a field on the `Ident`, we'll have to make it empty/unknown. This isn't an issue in-and-of-itself, it just means that we won't get the typesystem's help to enforce whether or not an `Ident` has a location.

A second benefit to wrapping in `Located` is that we can write a bunch of functions that take `Located<T>` (e.g. error messages) and use the location (yes, this could be accomplished with a `Located` trait too). It's essentially like a smart pointer.

All of that said, I'm happy to consider another approach — I only have perspective from the one project I've been working on and want to ensure that the contributions are aligned with the long term of the community around this repo.

Reply:
> This isn't an issue in-and-of-itself, it just means that we won't get the typesystem's help to enforce whether or not an Ident has a location.


this makes sense -- I guess I am worried about the implications on downstream crates of changing all fields to `Location` -- specifically when used in `match` expressions there are lots of patterns like https://github.com/apache/arrow-datafusion/blob/350cb47289a76e579b221fe374e4cf09db332569/datafusion/sql/src/expr/function.rs#L168-L184

Adding `Located` will require all such locations to add a new level of Located` . If we add `location` as a field then at least where there are matches with default match arms  `..` no changes will be required. 

```rust
match sql {
  FunctionArgs::Named {
    name, 
    arg: ...
    .. // <-- can be used to ignore location
  }
   ...
```

with located wouldn't this end up looking something like

```rust
match sql {
  FunctionArgs::Named {
    name: Located { name, ... }, 
    arg: Located { ... }
    .. // <-- can be used to ignore location
  }
   ...
```

In terms of converting String to Ident what about something like an optional location? Like

```rust
pub struct Ident {
    pub value: String,
    pub quote_style: Option<char>,
    location: Option<Location>,
}
```

And then have 
```rust
impl FromStr for Ident { 
...
 Ident { 
 ... 
  location: None.
}
```


> (yes, this could be accomplished with a Located trait too).

I agree this would be a better approach

Reply:
> with located wouldn't this end up looking something like

```
match sql {
  FunctionArgs::Named {
    name: Located { name, ... }, 
    arg: Located { ... }
    .. // <-- can be used to ignore location
  }
   ...
```

I don't think so — but apologies if I'm missing something. `Located` is exactly like a smart pointer (e.g. like `Box`), so you can write

```
match sql {
  FunctionArgs::Named {
    name, 
    arg,
}
...
```

and then "pretend" `name` is just an `Ident`, except that you can also call `.location()` on it.


> In terms of converting String to Ident what about something like an optional location? Like

Yes that's true, but the problem is that now you don't have the typesystem's help to know whether or not something has a location. For example, imagine you put these idents into a map where you "don't care" about locations. You'd need `Hash`, `Eq`, etc. to ignore the `location` field. However, there may be a separate use case where you do care about locations — what do you do then?
  }
   ...
```

Reply:
Thank you for your thoughtful reply @ankrgyl  -- I plan to give this a try in the next few days. I am sorry I just don't have very much focus time to devote to sqlparser at the moment

This PR is definitely on my list

Reply:
Thank you! I actually open sourced the project I'm working on that uses `sqlparser-rs` this weekend: https://github.com/qscl/queryscript. It's not ready for real use yet, but you can see the [fork of sqlparser](https://github.com/qscl/sqlparser-rs), and if you feel compelled, how we use it in the code (e.g. the VSCode uses Locations to show errors in the exact right locations). Happy to chat further about it offline too if helpful.

Reply:
I am sorry for my lack of communication here.

Basically I have (obviously) extremely limited time for maintenance of this crate and for that I apologize.

Given the limited maintenance bandwidth, I am very hesitant to accept a change that causes downstream churn given the possibility of requiring non trivial effort to fix here. 

Basically, to merge a PR like this I would like to see what changes are required to an existing project that uses AST matching  extensively. Any project that uses sqlparser-rs would do (though I am personally very familiar with  https://github.com/apache/arrow-datafusion)

Reply:
So specifically, I want to see a PR to an existing project that pins sqlparser to this branch and shows the tests / CI passing as a demonstration of how much downstream impact it would have)